logserver.c

```diff
@@ -347,127 +372,350 @@ static int logserver_handle_msg(struct logserver *logserver,
+	return 0;
+}
+
+static struct logserver_fd *logserver_fd_from_list(struct dl_list *l, int fd)
 {
+	struct logserver_fd *it, *tmp;
+	dl_list_for_each_safe(it, tmp, l, struct logserver_fd, list)
+	{
+		if (it->fd == fd) {
+			return it;
+		}
+	}
+	return NULL;
+}
+
+static struct logserver_fd *logserver_get_fd(int sockfd)
+{
+	union {
+		char buf[CMSG_SPACE(sizeof(int))];
+		struct cmsghdr align;
+	} ctrl;
+
+	char platform[LOGSERVER_MAX_HEADER_LEN] = { 0 };
+	char src[LOGSERVER_MAX_HEADER_LEN] = { 0 };
+
+	struct iovec iov[2];
+	iov[0] = (struct iovec){ .iov_base = platform,
+				 .iov_len = LOGSERVER_MAX_HEADER_LEN };
+	iov[1] = (struct iovec){ .iov_base = src,
+				 .iov_len = LOGSERVER_MAX_HEADER_LEN };
+
+	struct msghdr msg = { .msg_name = NULL,
+			      .msg_namelen = 0,
+			      .msg_iov = iov,
+			      .msg_iovlen = 2,
+			      .msg_control = ctrl.buf,
+			      .msg_controllen = sizeof(ctrl.buf) };
+
+	errno = 0;
+	if (recvmsg(sockfd, &msg, 0) < 0) {
+		pv_log(ERROR, "error receiving fd: %s", strerror(errno));
+		return NULL;
+	}
+
+	struct cmsghdr *cmsg = CMSG_FIRSTHDR(&msg);
+	if (!cmsg) {
+		pv_log(ERROR, "error receiving fd, NULL structure\n");
+		return NULL;
+	}
+
 	int fd;
-	struct sockaddr_un addr;
+	memcpy(&fd, CMSG_DATA(cmsg), sizeof(int));
+
+	return logserver_fd_new(platform, src, fd);
+}
+
+static int logserver_epoll_wait(struct epoll_event *ev)
+{
+	int ready = 0;
+	errno = 0;
+	do {
+		ready = epoll_wait(logserver_g.epfd, ev, LOGSERVER_MAX_EV, -1);
+
+		if (errno != 0 && errno != EINTR) {
+			pv_log(ERROR, "error calling epoll_wait: %s",
+			       strerror(errno));
+			return 0;
+		}
+	} while (ready < 0);
+
+	return ready;
+}
+
+static void logserver_consume_data(int fd, bool sub)
+{
+	struct buffer *buffer = pv_buffer_get(true);
+	ssize_t size = 0;
+	if (buffer) {
+		size = pv_fs_file_read_nointr(fd, buffer->buf, buffer->size);
+
+		if (size < 0)
+			goto out;
+	}
+
+	// if fd has name, so is a subcribed fd
+	if (sub) {
+		struct logserver_fd *lfd =
+			logserver_fd_from_list(&logserver_g.fdlst, fd);
+
+		struct logserver_msg_data d = { .version = LOG_PROTOCOL_LEGACY,
+						.level = DEBUG,
+						.tsec = (uint64_t)time(NULL),
+						.platform = lfd->platform,
+						.source = lfd->src,
+						.data = buffer->buf,
+						.data_len = size };
+		logserver_log_msg_data(&d);
+
+	} else {
+		struct logserver_msg *msg = (struct logserver_msg *)buffer->buf;
+		logserver_handle_msg(msg);
+	}
+out:
+	pv_buffer_drop(buffer);
+}
+
+static void logserver_loop()
+{
+	struct epoll_event ev[LOGSERVER_MAX_EV];
+	int n_events = logserver_epoll_wait(ev);
+
+	if (n_events < 1) {
+		return;
+	}
+
+	int logsock = logserver_g.logsock;
+	int fdsock = logserver_g.fdsock;
+	struct dl_list *tmplst = &logserver_g.tmplst;
+	struct dl_list *fdlst = &logserver_g.fdlst;
+
+	int curfd = -1;
+	for (int i = 0; i < n_events; ++i) {
+		curfd = ev[i].data.fd;
+		if (curfd == logsock || curfd == fdsock) {
+			int fd = logserver_accept_connection(curfd);
```

Comment:
maybe do this in different functions (the content of this if, then the content of the else if, then the else)? this kind of code creates a lot of indentation and very big functions like this one. Not necessarily bad but hard to read

Reply:
Yeah, that is on my backlog ;)